lib/Trace.js

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2015-2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+class Node {
+    constructor(span, type) {
+        this._span = span;
+        this._type = type;
+        this._nodes = [];
+    }
+
+    getSpan() {
+        return this._span;
+    }
+
+    setType(type) {
+        this._type = type;
+    }
+
+    getNodes() {
+        return this._nodes;
+    }
+
+    toJSON() {
+        const duration = this._span.getEndTime() && this._span.getStartTime() ?
+            this._span.getEndTime() - this._span.getStartTime() : 0;
+
+        const node = {
+            type: this._type,
+            // TODO http.uri http.path?
+            uri: Node._deriveUrl(this._span.getTags()),
+            operation: Node._deriveOperation(this._span),
```

Comment:
Both of these functions (`_deriveOperation` and `_deriveUrl`) end up as semi-private static methods on the class. One way to avoid this, is to either declare the functions here in `toJSON()`, or more efficiently, just declare them as functions outside of the class altogether, effectively hiding them from client view.

This also applies to `_tagsToProperties` below.